### PR TITLE
*: Support stream aggregation in new plan

### PR DESCRIPTION
**Pull request by zimulala**
support stream aggregation in the new plan and clean the tests.

### CODE REVIEW

File: plan/cbo_test.go [hunk 1]
```diff
@@ -70,19 +72,43 @@ func (s *testAnalyzeSuite) TestIndexRead(c *C) {
...
+		{
+			sql:  "select count(*) from t where c > '1' group by b",
+			best: "IndexReader(Index(t.b_c)[[<nil> <nil>,+inf +inf]]->Sel([gt(test.t.c, 1)]))->StreamAgg",
+		},
+		{
+			sql:  "select count(*) from t where e = 1 group by b",
+			best: "IndexLookUp(Index(t.e)[[1,1]], Table(t)->HashAgg)->HashAgg",
+		},
+		{
+			sql:  "select count(*) from t where e > 1 group by b",
```
Comment (hanfei1991) [inline plan/cbo_test.go]: the condition `e>1`'s filter rate is 99.98%. Why should we push agg down ?

File: plan/dag_plan_test.go [hunk 2]
```diff
@@ -380,7 +380,7 @@ func (s *testPlanSuite) TestDAGPlanBuilderSubquery(c *C) {
 		// Test Nested sub query.
 		{
```
Comment (hanfei1991) [inline plan/dag_plan_test.go]: add test for stream agg + index join and stream agg + merge join and stream agg + limit and stream agg + sort

File: plan/logical_plans.go [hunk 3]
```diff
@@ -179,11 +179,12 @@ type LogicalAggregation struct {
 
 	AggFuncs     []expression.AggregationFunction
 	GroupByItems []expression.Expression
-
 	// groupByCols stores the columns that are group-by items.
 	groupByCols []*expression.Column
 
 	possibleProperties [][]*expression.Column
+	childCount         float64 // childCount is the child plan's count.
+	cardinality        float64
```
Comment (hanfei1991) [inline plan/logical_plans.go]: cardinality is always equal to LogicalAggregation's profile's count.

File: plan/logical_plans.go [hunk 4]
```diff
@@ -179,11 +179,12 @@ type LogicalAggregation struct {
 
 	AggFuncs     []expression.AggregationFunction
 	GroupByItems []expression.Expression
-
 	// groupByCols stores the columns that are group-by items.
 	groupByCols []*expression.Column
 
 	possibleProperties [][]*expression.Column
+	childCount         float64 // childCount is the child plan's count.
```
Comment (hanfei1991) [inline plan/logical_plans.go]: maybe `inputCount` is a better name. childCount seems to indicate the number of children.

File: plan/new_physical_plan_builder.go [hunk 5]
```diff
@@ -1090,26 +1090,80 @@ func (p *TopN) getChildrenPossibleProps(prop *requiredProp) [][]*requiredProp {
 	return props
 }
 
+func (p *LogicalAggregation) getStreamAggs() []PhysicalPlan {
+	for _, aggFunc := range p.AggFuncs {
+		if aggFunc.GetMode() == expression.FinalMode {
```
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: This code seems never be touched ?

File: plan/new_physical_plan_builder.go [hunk 6]
```diff
@@ -1090,26 +1090,80 @@ func (p *TopN) getChildrenPossibleProps(prop *requiredProp) [][]*requiredProp {
...
+			return nil
+		}
+	}
+	// group by a + b is not interested in any order.
```
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: Add test for this case.

File: plan/task.go [hunk 7]
```diff
@@ -490,19 +490,22 @@ func (p *PhysicalAggregation) attach2Task(tasks ...task) task {
...
 				cop.tablePlan = partialAgg
-				cop.cst += cop.count() * cpuFactor
 			} else {
 				partialAgg.SetChildren(cop.indexPlan)
 				cop.indexPlan = partialAgg
-				cop.cst += cop.count() * cpuFactor
 			}
 		}
 		task = finishCopTask(cop, p.ctx, p.allocator)
+		task.addCost(task.count()*cpuFactor + p.cardinality*hashAggMemFactor)
```
Comment (hanfei1991) [inline plan/task.go]: use p.statsProfile().count to replace `p.cardinality`

File: plan/new_physical_plan_builder.go (hunk 5 above)
Comment (zimulala) [reply plan/new_physical_plan_builder.go]: We do `decompose` will have `FinalMode` aggregation.

File: plan/stats.go [hunk 8]
```diff
@@ -81,7 +81,6 @@ func (p *DataSource) getStatsProfileByFilter(conds expression.CNFExprs) *statsPr
 	}
 	selectivity, err := p.statisticTable.Selectivity(p.ctx, conds)
 	if err != nil {
-		log.Warnf("An error happened: %v, we have to use the default selectivity", err.Error())
```
Comment (hanfei1991) [inline plan/stats.go]: Why remove this log ?

File: plan/new_physical_plan_builder.go [hunk 9]
```diff
@@ -1090,26 +1090,80 @@ func (p *TopN) getChildrenPossibleProps(prop *requiredProp) [][]*requiredProp {
...
+		}
+		return props
 	}
-	props := make([][]*requiredProp, 0, len(wholeTaskTypes))
-	for _, tp := range wholeTaskTypes {
-		props = append(props, []*requiredProp{{taskTp: tp, expectedCnt: math.MaxFloat64}})
+
+	reqProp := &requiredProp{taskTp: rootTaskType, cols: p.propKeys, expectedCnt: prop.expectedCnt * p.childCount / p.profile.count}
+	if !prop.isEmpty() {
+		if prop.desc {
```
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: Desc can also pass

File: plan/new_physical_plan_builder.go [hunk 10]
```diff
@@ -1090,26 +1090,80 @@ func (p *TopN) getChildrenPossibleProps(prop *requiredProp) [][]*requiredProp {
...
 		AggType:      CompleteAgg,
+		cardinality:  p.cardinality,
 	}.init(p.allocator, p.ctx)
-	ha.SetSchema(p.schema)
-	ha.profile = p.profile
-	return []PhysicalPlan{ha}
+	agg.SetSchema(p.schema.Clone())
+	agg.profile = p.profile
+	aggs = append(aggs, agg)
+	if len(p.possibleProperties) == 0 {
```
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: we can remove this check.

File: plan/new_physical_plan_builder.go [hunk 11]
```diff
@@ -1090,26 +1090,80 @@ func (p *TopN) getChildrenPossibleProps(prop *requiredProp) [][]*requiredProp {
...
+		agg.SetSchema(p.schema.Clone())
+		agg.profile = p.profile
+		streamAggs = append(streamAggs, agg)
+	}
+	return streamAggs
+}
+
 func (p *LogicalAggregation) generatePhysicalPlans() []PhysicalPlan {
-	ha := PhysicalAggregation{
+	aggs := make([]PhysicalPlan, 0, 2)
```
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: 2 is not reasonable, len(p.possibleProperties) + 1 is the upper bound.

**Comment by shenli** [comment]
@hanfei1991 PTAL

File: plan/cbo_test.go [hunk 12]
```diff
@@ -70,19 +72,43 @@ func (s *testAnalyzeSuite) TestIndexRead(c *C) {
...
+			best: "TableReader(Table(t)->Sel([gt(test.t.e, 1)])->HashAgg)->HashAgg",
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply plan/cbo_test.go]: using index will cause a double read.

File: plan/dag_plan_test.go (hunk 2 above)
Comment (zimulala) [reply plan/dag_plan_test.go]: Done.

File: plan/new_physical_plan_builder.go (hunk 6 above)
Comment (zimulala) [reply plan/new_physical_plan_builder.go]: Done.

**Comment by hanfei1991** [comment]
LGTM

File: plan/cbo_test.go [hunk 13]
```diff
@@ -70,19 +72,43 @@ func (s *testAnalyzeSuite) TestIndexRead(c *C) {
...
+		},
+		{
+			sql:  "select count(e) from t where t.b <= 20",
+			best: "IndexLookUp(Index(t.b_c)[[-inf <nil>,20 +inf]], Table(t)->HashAgg)->HashAgg",
+		},
+		{
+			sql:  "select count(e) from t where t.b <= 30",
```
[the remark this answers is not in the document]
Comment (zimulala) [reply plan/cbo_test.go]: There are old tests. @hanfei1991

[the remark this answers is not in the document]
Comment (hanfei1991) [reply plan/cbo_test.go]: oh

**Comment by zimulala** [comment]
PTAL @winoros

**Pull request merged**